src/hooks/useImageSize.ts

```diff
@@ -19,138 +19,140 @@ import {
-const filterGreater = (
-  sizes: SizeArray,
-  size: SquareSizes | WidthSizes
-): SizeArray => {
-  return sizes.filter(s => getWidth(s) > getWidth(size))
-}
-
-/** Gets all sizes in an array that are less than the provided size */
-const filterLess = (
-  sizes: SizeArray,
-  size: SquareSizes | WidthSizes
-): SizeArray => {
-  return sizes.filter(s => getWidth(s) < getWidth(size))
-}
+const getWidth = (size: Size): number => parseInt(size.split('x')[0], 10)
 
 /** Sorts sizes according to their width dimension */
-const sortSizes = (sizes: SizeArray, reverse?: boolean): SizeArray => {
-  return reverse
-    ? sizes.sort((a, b) => getWidth(b) - getWidth(a))
-    : sizes.sort((a, b) => getWidth(a) - getWidth(b))
+const sortSizes = <ImageSize extends Size>(sizes: ImageSize[]): ImageSize[] => {
+  return sizes.sort((a, b) => getWidth(b) - getWidth(a))
 }
 
 /**
- * Gets the next smallest available image size. If we have images
- * [A > B > C] and we request B, this method returns C.
+ * Gets the next available image size (sorted largest to smallest) meeting a condition
  */
-const getNextSmallest = (
-  imageSizes: ImageSizes,
-  size: SquareSizes | WidthSizes,
-  defaultImage: string
-): URL => {
-  const keys = Object.keys(imageSizes) as SquareSizes[]
-  const nextLargest = sortSizes(filterLess(keys, size))[0]
-  if (!nextLargest) return defaultImage
-  return (imageSizes as any)[nextLargest]
+const getNextImage = <
+  ImageSize extends Size,
+  ImageSizes extends ImageSizesObject<ImageSize>
+>(
+  condition: (desiredWidth: number, currentWidth: number) => boolean
+) => (imageSizes: ImageSizes, size: ImageSize) => {
+  const keys = Object.keys(imageSizes) as ImageSize[]
+
+  const desiredWidth = getWidth(size)
+  const sorted = sortSizes(
+    keys.filter(s => condition(getWidth(s), desiredWidth))
+  )
+  const next = sorted[0]
+  return imageSizes[next]
 }
 
-/**
- * Gets the first available larger image size. If we have images
- * [A > B > C] and we request C, this method returns A.
- */
-const getLarger = (
-  imageSizes: ImageSizes,
-  size: SquareSizes | WidthSizes,
-  defaultImage: string
-): URL | null => {
-  const keys = Object.keys(imageSizes) as SquareSizes[]
-  const larger = sortSizes(filterGreater(keys, size))[0]
-  if (!larger) return defaultImage
-  return (imageSizes as any)[larger]
-}
-
-type UseImageSizeProps = {
-  id?: number | null
-  sizes: ImageSizes | null
-  size: SquareSizes | WidthSizes
-  action: (id: number, size: SquareSizes | WidthSizes) => void
+type UseImageSizeProps<
+  ImageSize extends Size,
+  ImageSizes extends ImageSizesObject<ImageSize>
+> = {
+  // The action to dispatch to fetch the desired size
+  action: (id: number, size: ImageSize) => void
+  // The fallback if no sizes are available
   defaultImage?: string
+  // A unique id (used to prevent duplicate fetches)
+  id?: number | null
+  // A flag (default = true) that will trigger the image load. Can be used to delay the load
+  load?: boolean
```

Comment:
This is a new option because `onDemand` doesn't allow for using an already cached image before triggering the fetch for the desired size